src/Nethermind/Nethermind.Era1/E2Store.cs

```diff
@@ -0,0 +1,281 @@
+
+    }
+    public Task<int> WriteEntry(UInt16 type, byte[] bytes, CancellationToken cancellation = default)
+    {
+        return WriteEntry(type, bytes, false, cancellation);
+    }
+
+    private async Task<int> WriteEntry(UInt16 type, byte[] bytes, bool asSnappy, CancellationToken cancellation = default)
+    {
+        //TODO refactor to sync and use span?
+        byte[] headerBuffer = ArrayPool<byte>.Shared.Rent(HeaderSize);
+        try
+        {
+            //See https://github.com/google/snappy/blob/main/framing_format.txt
+            if (asSnappy && bytes.Length > 0)
+            {
+                //TODO find a way to write directly to file, and still return the number of bytes written
+                using MemoryStream memoryStream = new MemoryStream();
+                using SnappyStream snappyStream = new(memoryStream, CompressionMode.Compress, true);
+                await snappyStream.WriteAsync(bytes, cancellation);
+                await snappyStream.FlushAsync();
+                bytes = memoryStream.ToArray();
+            }
+
+            headerBuffer[0] = (byte)type;
+            headerBuffer[1] = (byte)(type >> 8);
+            int length = bytes.Length;
+            headerBuffer[2] = (byte)(length);
+            headerBuffer[3] = (byte)(length >> 8);
+            headerBuffer[4] = (byte)(length >> 16);
+            headerBuffer[5] = (byte)(length >> 24);
+            headerBuffer[6] = 0;
+            headerBuffer[7] = 0;
+
+            await _stream.WriteAsync(headerBuffer, 0, HeaderSize, cancellation);
+            if (length > 0) await _stream.WriteAsync(bytes, 0, length, cancellation);
+
+            return length + HeaderSize;
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(headerBuffer);
+        }
+    }
+
+    private async Task<EraMetadata> ReadEraMetaData(CancellationToken token = default)
+    {
+        long l = _stream!.Length;
+        if (_stream.Length < 16)
+            throw new EraFormatException($"Data is not in a valid Era format.");
+
+        byte[] bytes = ArrayPool<byte>.Shared.Rent(16);
+        try
+        {
+            _stream.Position = _stream.Length - 8;
+            await _stream.ReadAsync(bytes, 0, 8, token);
+            long c = BitConverter.ToInt64(bytes);
+
+            long indexOffset = l - 16L - c * 8;
+            _stream.Position = indexOffset;
+            await _stream.ReadAsync(bytes, 8, 8, token);
+            long s = BitConverter.ToInt64(bytes, 8);
+            return new EraMetadata(s, c, l);
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(bytes);
+        }
+    }
+
+    // Reads the header metadata at the given offset.
+    public async Task<HeaderData> ReadEntryHeaderAt(long offset, CancellationToken token = default)
+    {
+        CheckStreamBounds(offset);
+
+        var buf = ArrayPool<byte>.Shared.Rent(HeaderSize);
+        try
+        {
+            _stream!.Position = offset;
```

Comment:
Are we doing a lot of jumping/seeking on the stream or is it more of linear scans?

Reply:
Every iteration will read the offset from the block index, which is appended to the file. 
That will probably trigger a lot of random reads. 
I thought about loading the whole index instead at the expense of memory WDYT?